File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys

input = sys.stdin.readline

cnt = int(input())
stack = []
answer = []
num = 1

for _ in range(cnt):
    n = int(input())
    if answer and answer[0] == "NO":
        continue
    if n == num:
        answer.append("+")
        answer.append("-")
        num += 1
        continue
    if not stack:
        stack.append(num)
        answer.append("+")
        num += 1
    while stack[-1] != n:
        if num > cnt:
            break
        stack.append(num)
        answer.append("+")
        num += 1
    if stack[-1] == n:
        stack.pop()
        answer.append("-")
    else:
        answer = ["NO"]

for i in answer:
    print(i)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());
        Stack<Integer> arr = new Stack<>();
        int flag = 1;
        int num = Integer.parseInt(br.readLine());

        for (int i = 1; i < n + 1; i++){
            arr.push(i);
            sb.append("+\n");

            while (!arr.isEmpty()){
                int temp = arr.pop();
                if (flag == n){
                    break;
                }
                if (temp == num){
                    num = Integer.parseInt(br.readLine());
                    sb.append("-\n");
                    flag+=1;
                }else{
                    arr.push(temp);
                    break;
                }
            }
        }
        for(int i = flag; i < n; i++) br.readLine();
        if (arr.isEmpty()){
            System.out.println(sb.append("-"));
        }else{
            System.out.println("NO");
        }

    }
}
